packages/server/src/integrations/base/clientCache.ts

```diff
@@ -0,0 +1,65 @@
+import { Duration } from "@budibase/backend-core"
+import LRUCache from "lru-cache"
+
+interface Merkle {
+  hash: (json: Record<string, any>) => string
+  stringify: (json: Record<string, any>) => string
+}
+
+interface MerkleConstructor {
+  new (): Merkle
+}
+
+// no typescript definition - fill it in
+const { MerkleJson } = require("merkle-json") as {
```

Comment:
`merkle-json` appears to cache the hash values of objects on the object itself under the key `merkleHash`. Does this open up the possibility of users putting `merkleHash` on their DS config objects to try and "steal" a connection to another user's datasource? Should we add something that isolates tenants from each other explicitly instead of relying on the hash being unique?

Reply:
The merkle-json library only does this if we call like `merkle.hash(json, true)` - the boolean parameter telling it to re-use hashes - although I'm not a huge fan of this library due to its lack of in-built TS support and its relatively low usage, there is probably a better way to do this.

I did consider using the `datasourceId` instead for the key, but the downside to this approach is connections are cached even if the config has changed which would lead to weird behaviour. We could retrieve the tenantId and add it to the start of the hash though, like `tenantId-hash` for the cache key.

Reply:
I don't think it'd be terribly challenging to do the hashing ourselves. Concat the tenantId, host, port, user, password, whatever else, do a sha256 on that.